### ISSUE TITLE
Serializing a struct and its methods for later reuse

### ISSUE THREAD

**Issue by ChrisTrenkamp** [gopherjs/gopherjs#828]
[Here's the backstory.](https://github.com/blevesearch/bleve/issues/667)

Let's say you have a static web site and you want to provide a search via a pre-populated search index.  So you write a little program to read in the content of your site and create some search indices.  We can use Bleve and utilize its in-memory store since it doesn't rely on os-specific functionality like mmap.

```go
package main

import "github.com/blevesearch/bleve"

func createIndex(idx bleve.Index) {
	//Some code for walking the website and populating the index...
}

func main() {
	index, _ := bleve.NewMemOnly(bleve.NewIndexMapping())
	createIndex(index)
	// The Bleve store is now populated with your sites search indices...
	// ... now what?
}
```

Given the above pseudocode, being ran with GopherJS, to create the search indices, is there a way to serialize it to a file so it can be pulled in by the website?  The above link shows ways to use GopherJS to build an index using GopherJS, but I can't figure out how to use GopherJS to generate a pre-built search index.

Any help will be appreciated.

**Comment by myitcv** [comment]
If I've understood correctly (and please correct where I go wrong, I've never used Bleve in anger) you're after:

* the ability to pre-populate a Bleve index and persist that index to some format
* make the persisted data available in a file such that a frontend (browser) can download the persisted index
* have your frontend single-page-application/page load a GopherJS-based application script, which in turn reifies the index from a data URL in such as way as to make the index available to the application/page

I think the question therefore becomes: which of the existing Bleve indexes:

* support being compiled by GopherJS (anything that requires disk access, as you've pointed out, is off the cards here)
* have something like `io.Reader`/`io.Writer` interface for their persistence, which would allow using an `io.Reader` (URL) to reify the index

@mschoch might be able to help answer those questions?

Two alternatives I can think of:

1. Write a [localStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage) index store "driver" - I'm not even sure whether the API of local storage is sufficiently rich for Bleve, and/or whether some aspects would need to be faked. This would then require you to serve a file that gets loaded in to localStorage (when it changes, taking advantage of caching), but your GopherJS application would then at least be able to use a Bleve index in the usual way (backed by the localStorage index store)
2. Code generate (part of) a GopherJS application that populates a mem-only index and makes it available (via some API) on `window`

There are different costs to all these options:

* size of compiled GopherJS result
* size of persisted index data
* CPU/memory cost in the browser
* implementation cost

Honestly without comparing I can't give you a strong steer on the CPU/memory/transfer pros/cons of each. And I'm rather blind on the cost of implementing a localStorage index store too.

In implementation cost terms, the code-generation solution (point 2 above) is something you could do now and getting running pretty quickly. Using `gopherjs build -m` (minify) you will at least optimise the transfer cost.

**Comment by ChrisTrenkamp** [comment]
That's correct.  localStorage is a no-go as well because the size of the index can get quite big.  In-memory will be fine.

What I would ideally like is to serialize the Bleve index to plain Javascript.  That way, the browser just has to parse the Javascript file without going through an extra step to load the indices.  It'll also make it easier for the programmer; to get the site's index, just pull in that Javascript file.  I know the file size will be bloated, but file size is not my concern.  My users actually view these sites directly off the disk instead of through a web server (long story).

I think what I'm looking for is something like the `encoding/json` package, except it'll serialize unexported fields as well.

**Comment by ChrisTrenkamp** [comment]
The data structure is a binary search tree.  [https://github.com/steveyen/gtreap](https://github.com/steveyen/gtreap).

**Comment by myitcv** [comment]
Ok, so you're going to have to handle reifying the pointers etc too, right?

This was the reason behind asking whether any of the existing index store types support persisting to/reading from file.

**Comment by ChrisTrenkamp** [comment]
Correct.  It would be great if there was a utility that could handle pointers, recursion, etc.  I'm trying to avoid an extra parse phase because I want the index to be available ASAP.  I've looked at projects like [this](https://github.com/WebReflection/circular-json) one, but it's still requires a call to parse the data.

**Comment by ChrisTrenkamp** [comment]
I tested loading the search indices from a line-delimited file.  It's taking 30x longer to load the indices compared to native.  Removing gzip compression shaved off some time, but it still has to deal with decoding base64 and rebalancing the tree structure while loading the file.
